File: js/moq/src/ietf/track.ts
import type * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Namespace from "./namespace.ts";

export class TrackStatusRequest {
	static id = 0x0d;

	trackNamespace: Path.Valid;
	trackName: string;

	constructor(trackNamespace: Path.Valid, trackName: string) {
		this.trackNamespace = trackNamespace;
		this.trackName = trackName;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.trackNamespace);
		await w.string(this.trackName);
	}

	static async decodeMessage(r: Reader): Promise<TrackStatusRequest> {
		const trackNamespace = await Namespace.decode(r);
		const trackName = await r.string();
		return new TrackStatusRequest(trackNamespace, trackName);
	}
}

// Track status message for communicating track-level state
export class TrackStatus {
	static id = 0x0e;

	trackNamespace: Path.Valid;
	trackName: string;
	statusCode: number;
	lastGroupId: bigint;
	lastObjectId: bigint;

	constructor(
		trackNamespace: Path.Valid,
		trackName: string,
		statusCode: number,
		lastGroupId: bigint,
		lastObjectId: bigint,
	) {
		this.trackNamespace = trackNamespace;
		this.trackName = trackName;
		this.statusCode = statusCode;
		this.lastGroupId = lastGroupId;
		this.lastObjectId = lastObjectId;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.trackNamespace);
		await w.string(this.trackName);
		await w.u62(BigInt(this.statusCode));
		await w.u62(this.lastGroupId);
		await w.u62(this.lastObjectId);
	}

	static async decodeMessage(r: Reader): Promise<TrackStatus> {
		const trackNamespace = await Namespace.decode(r);
		const trackName = await r.string();
		const statusCode = Number(await r.u62());
		const lastGroupId = await r.u62();
		const lastObjectId = await r.u62();

		return new TrackStatus(trackNamespace, trackName, statusCode, lastGroupId, lastObjectId);
	}

	// Track status codes
	static readonly STATUS_IN_PROGRESS = 0x00;
	static readonly STATUS_NOT_FOUND = 0x01;
	static readonly STATUS_NOT_AUTHORIZED = 0x02;
	static readonly STATUS_ENDED = 0x03;
}

File: rs/moq/src/ietf/track.rs
//! IETF moq-transport-07 track status messages

use std::borrow::Cow;

use crate::{coding::*, Path};

use super::util::{decode_namespace, encode_namespace};

/// TrackStatusRequest message (0x0d)
#[derive(Clone, Debug)]
pub struct TrackStatusRequest<'a> {
	pub track_namespace: Path<'a>,
	pub track_name: Cow<'a, str>,
}

impl<'a> Message for TrackStatusRequest<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.track_namespace);
		self.track_name.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let track_namespace = decode_namespace(r)?;
		let track_name = Cow::<str>::decode(r)?;

		Ok(Self {
			track_namespace,
			track_name,
		})
	}
}

/// TrackStatus message (0x0e)
/// Sent to communicate track-level state
#[derive(Clone, Debug)]
pub struct TrackStatus<'a> {
	pub track_namespace: Path<'a>,
	pub track_name: Cow<'a, str>,
	pub status_code: u64,
	pub last_group_id: u64,
	pub last_object_id: u64,
}

impl<'a> TrackStatus<'a> {
	pub const STATUS_IN_PROGRESS: u64 = 0x00;
	pub const STATUS_NOT_FOUND: u64 = 0x01;
	pub const STATUS_NOT_AUTHORIZED: u64 = 0x02;
	pub const STATUS_ENDED: u64 = 0x03;
}

impl<'a> Message for TrackStatus<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.track_namespace);
		self.track_name.encode(w);
		self.status_code.encode(w);
		self.last_group_id.encode(w);
		self.last_object_id.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let track_namespace = decode_namespace(r)?;
		let track_name = Cow::<str>::decode(r)?;
		let status_code = u64::decode(r)?;
		let last_group_id = u64::decode(r)?;
		let last_object_id = u64::decode(r)?;

		Ok(Self {
			track_namespace,
			track_name,
			status_code,
			last_group_id,
			last_object_id,
		})
	}
}
